src/wdf.cpp

```diff
@@ -409,12 +416,32 @@ WDF_DRIVER_CONFIG_INIT(_Out_ PWDF_DRIVER_CONFIG config, _In_opt_ PFN_WDF_DRIVER_
     *config = {.EvtDriverDeviceAdd = evt_driver_device_add};
 }
 
-HANDLE
-usersim_get_device_handle(HMODULE module, _In_opt_z_ const WCHAR* device_name)
+WDFDRIVER
+usersim_get_driver_from_module(HMODULE module)
+{
+    usersim_dll_get_driver_from_module_t usersim_dll_get_driver_from_module =
+        (usersim_dll_get_driver_from_module_t)GetProcAddress(module, "usersim_dll_get_driver_from_module");
+    return usersim_dll_get_driver_from_module();
+}
+
+WDFDEVICE
+usersim_get_device_by_name(WDFDRIVER driver, _In_opt_ PCWSTR device_name)
 {
-    usersim_dll_get_device_handle_t usersim_dll_get_device_handle =
-        (usersim_dll_get_device_handle_t)GetProcAddress(module, "usersim_dll_get_device_handle");
-    return usersim_dll_get_device_handle(device_name);
+    PDRIVER_OBJECT driver_object = (PDRIVER_OBJECT)driver;
+    if (!device_name) {
+        return (!driver_object->devices.empty()) ? driver_object->devices.front() : nullptr;
+    }
+    UNICODE_STRING unicode_device_name;
+    RtlInitUnicodeString(&unicode_device_name, device_name);
+    for (PDEVICE_OBJECT device : driver_object->devices) {
+        if (device->init.device_name.Length != unicode_device_name.Length) {
```

Comment:
Worth creating an RtlEqualUnicodeString function?

Reply:
My view: only if there's more than 1 occurrence.  So far there's not.